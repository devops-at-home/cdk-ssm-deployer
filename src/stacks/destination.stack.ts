import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SSMRole, SSMRoleConfig } from '../constructs/ssm-role.construct';

interface DestinationStackProps extends SSMRoleConfig, StackProps {
    instanceName: string;
    environment: string;
}

export class DestinationStack extends Stack {
    public readonly roleName: string;
    constructor(scope: Construct, id: string, props: DestinationStackProps) {
        super(scope, id, props);

        const { roleName } = new SSMRole(this, 'Role', {
            ...props,
        });

        this.roleName = roleName;
    }
}
