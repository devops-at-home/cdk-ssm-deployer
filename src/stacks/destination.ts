import { NestedStack, NestedStackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SSMRole } from '../constructs/ssm-role';
// import { SSMDeployment } from '../constructs/ssm-deployment';

interface DestinationStackProps extends NestedStackProps {
  bucketName: string;
  tableName: string;
  destination: string;
  keyArn: string;
}

export class DestinationStack extends NestedStack {
  constructor(scope: Construct, id: string, props: DestinationStackProps) {
    super(scope, id, props);

    const { bucketName, tableName, destination, keyArn } = props;

    new SSMRole(this, 'Role', {
      destination,
      bucketName,
      tableName,
      keyArn,
    });

    // new SSMDeployment(this, 'Deployment', {
    //   bucketName,
    //   destination,
    // });
  }
}
