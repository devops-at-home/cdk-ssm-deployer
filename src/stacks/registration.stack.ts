import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { SSMRegistration, SSMRegistrationProps } from '../constructs/ssm-registration.construct';

interface RegistrationStackProps extends StackProps, SSMRegistrationProps {}

export class RegistrationStack extends Stack {
    constructor(scope: Construct, id: string, props: RegistrationStackProps) {
        super(scope, id, props);

        const { instanceName, roleName } = props;

        new SSMRegistration(this, 'SSMRegistration', {
            instanceName,
            roleName,
        });
    }
}
