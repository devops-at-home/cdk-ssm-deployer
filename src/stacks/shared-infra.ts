import { NestedStack, NestedStackProps } from 'aws-cdk-lib';
import { Table } from 'aws-cdk-lib/aws-dynamodb';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { BucketWithEventBridge } from '../constructs/config-bucket';
import { VaultDynamoDB } from '../constructs/vault-dynamodb';
import { GitHubActionsOidcProvider } from '../constructs/gh-aws-oidc-connect-provider';
import { GitHubActionsRole } from '../constructs/gh-aws-oidc-connect-role';
import { PolicyDocument, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { KmsKey } from '../constructs/kms-key';
import { Key } from 'aws-cdk-lib/aws-kms';

const repository = 'devops-at-home/cdk-ssm-deployer';
const destinations: string[] = [];

export class SharedInfraStack extends NestedStack {
  public bucket: Bucket;
  public table: Table;
  public key: Key;
  constructor(scope: Construct, id: string, props: NestedStackProps = {}) {
    super(scope, id, props);

    // TODO: SSM document for running deployments

    const { bucket } = new BucketWithEventBridge(this, 'Bucket');
    this.bucket = bucket;

    const { table } = new VaultDynamoDB(this, 'VaultDynamoDB');
    this.table = table;

    const { key } = new KmsKey(this, 'KmsKey');
    this.key = key;

    new GitHubActionsOidcProvider(this, 'GitHubActionsOidcProvider');

    new GitHubActionsRole(this, 'GitHubActionsRole', {
      provider: GitHubActionsOidcProvider.forAccount(),
      repository,
      inlinePolicies: {
        s3: new PolicyDocument({
          statements: [
            new PolicyStatement({
              resources: [`${bucket.bucketArn}/*`],
              actions: ['s3:PutObject'],
              conditions: {
                StringEquals: {
                  's3:prefix': destZips(destinations),
                },
              },
            }),
          ],
        }),
      },
    });
  }
}

const destZips = (destinations: string[]) => {
  return destinations.map((dest) => {
    return `${dest}.zip`;
  });
};
