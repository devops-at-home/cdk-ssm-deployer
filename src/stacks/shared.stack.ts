import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Bucket, BucketEncryption, IBucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

interface SharedStackProps extends StackProps {
    environment: string;
}

export class SharedStack extends Stack {
    public readonly bucket: IBucket;
    constructor(scope: Construct, id: string, props: SharedStackProps) {
        super(scope, id, props);

        const { environment } = props;

        const { bucketArn } = (this.bucket = new Bucket(this, 'Bucket', {
            encryption: BucketEncryption.S3_MANAGED,
        }));

        new CfnOutput(this, 'CfnOutput', {
            exportName: getBucketExportName(environment),
            value: bucketArn,
        });
    }
}

export const getBucketExportName = (environment: string) => {
    return `app-config-backup-bucket-${environment}`;
};
