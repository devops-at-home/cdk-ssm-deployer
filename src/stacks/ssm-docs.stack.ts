import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';

interface SSMDocsStackProps extends StackProps {}

export class SSMDocsStack extends Stack {
    constructor(scope: Construct, id: string, props: SSMDocsStackProps) {
        super(scope, id, props);

        // TODO: SSM document for running deployments
    }
}
