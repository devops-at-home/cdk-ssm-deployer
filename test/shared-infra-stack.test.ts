import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { SharedInfraStack } from '../src/stacks/shared-infra';

describe('SharedInfraStack', () => {
  const stack = new Stack();
  const nestedStack = new SharedInfraStack(stack, 'SharedInfraStack');
  const template = Template.fromStack(nestedStack);

  test('Bucket', () => {
    template.resourceCountIs('AWS::S3::Bucket', 1);
    template.hasResourceProperties('AWS::S3::Bucket', {
      EventBridgeConfiguration: { EventBridgeEnabled: 'true' },
      VersioningConfiguration: {
        Status: 'Enabled',
      },
      BucketEncryption: {
        ServerSideEncryptionConfiguration: [
          {
            ServerSideEncryptionByDefault: {
              SSEAlgorithm: 'AES256',
            },
          },
        ],
      },
    });
  });

  test('Snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
