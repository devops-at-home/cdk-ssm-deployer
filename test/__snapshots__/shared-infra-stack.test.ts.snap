// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SharedInfraStack Snapshot 1`] = `
Object {
  "Outputs": Object {
    "BucketConfigBucket6D3E4CBA": Object {
      "Value": Object {
        "Ref": "BucketD7FEB781",
      },
    },
    "KmsKeyOutputB5767BE1": Object {
      "Value": Object {
        "Ref": "KmsKeyBFFB7B85",
      },
    },
    "VaultDynamoDBVaultTable80F08FA5": Object {
      "Value": Object {
        "Ref": "VaultDynamoDBTable7271446C",
      },
    },
  },
  "Resources": Object {
    "BucketD7FEB781": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "NoncurrentVersionExpirationInDays": 60,
              "Status": "Enabled",
            },
          ],
        },
        "NotificationConfiguration": Object {
          "EventBridgeConfiguration": Object {
            "EventBridgeEnabled": "true",
          },
        },
        "VersioningConfiguration": Object {
          "Status": "Enabled",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "GitHubActionsOidcProviderGithubOidcProvider": Object {
      "Properties": Object {
        "ClientIdList": Array [
          "https://github.com/devops-at-home",
        ],
        "ThumbprintList": Array [
          "6938fd4d98bab03faadb97b34396831e3780aea1",
        ],
        "Url": "https://token.actions.githubusercontent.com",
      },
      "Type": "AWS::IAM::OIDCProvider",
    },
    "GitHubActionsRole4F1BBA26": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "StringLike": Object {
                  "token.actions.githubusercontent.com:sub": "repo:devops-at-home/*:ref:refs/heads/main*",
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:aws:iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":oidc-provider/token.actions.githubusercontent.com",
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": "s3:PutObject",
                  "Condition": Object {
                    "StringEquals": Object {
                      "s3:prefix": Array [
                        "h6060-001.zip",
                      ],
                    },
                  },
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        Object {
                          "Fn::GetAtt": Array [
                            "BucketD7FEB781",
                            "Arn",
                          ],
                        },
                        "/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "s3",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "KmsKeyBFFB7B85": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "VaultDynamoDBTable7271446C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "Path",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "Key",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "Path",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "Key",
            "KeyType": "RANGE",
          },
        ],
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
